### PR TITLE
Update temporal client metrics scope.

### PR DESCRIPTION
**Pull request by nishkrishnan**
This PR does a few things:
* Creates a custom metrics interceptor for each temporal client interface method (under our app stats scope)
* Modifies the temporal clients scope to be a generic `temporalclient`. The idea here is that this allows us to define a generic dashboard for all clients. The only thing that will differ is what namespace is used which is available in the metrics themselves.
* This also slightly changes how our scopes are defined in order to pass the reporter up the chain.
* Finally this adds support for closing the client which was missing before.

### CODE REVIEW

File: server/neptune/temporal/client.go [hunk 1]
```diff
@@ -1,26 +1,60 @@
...
+	}
+
+	// default to app scope unless specified otherwise
+	var clientScope tally.Scope
+	var clientScopeCloser io.Closer
+	if options.StatsReporter != nil {
+		clientScope, clientScopeCloser = tally.NewRootScope(tally.ScopeOptions{
+			Prefix:   StatsNamespace,
+			Reporter: options.StatsReporter,
+		}, time.Second)
```
Comment (samrabelachew) [inline server/neptune/temporal/client.go:53]: do we want to log/error if stats reporter was nil?

**Review by samrabelachew — DISMISSED** [review]

File: server/neptune/temporal/client.go [hunk 2]
```diff
@@ -39,5 +73,154 @@ func NewClient(scope tally.Scope, logger logur.Logger, cfg valid.Temporal) (clie
...
+	s := i.scope.SubScope("execute_workflow")
+
+	timer := s.Timer("latency").Start()
+	defer timer.Stop()
+
+	run, err := i.ClientOutboundInterceptorBase.Next.ExecuteWorkflow(ctx, in)
+
+	if err != nil {
+		s.Counter("error").Inc(1)
+		return run, err
```
Comment (Aayyush) [inline server/neptune/temporal/client.go:141]: nit: we could eliminate 2 return statements by wrapping the counter increment logic in an if-else block and do the same for other methods too? What do you think?

**Review by Aayyush — DISMISSED** [review]
left one small comment!

File: server/neptune/temporal/client.go (hunk 1 above)
Comment (nishkrishnan) [reply server/neptune/temporal/client.go:53]: nah im kind of thinking of this package now as a standalone library that we might let others use.  Seems fair to have metrics as optional if for example a service doesn't use tally.

File: server/neptune/temporal/client.go (hunk 2 above)
Comment (nishkrishnan) [reply server/neptune/temporal/client.go:141]: it's a pretty marginal improvement and I don't like elses haha so I'm going to leave as is.

Comment (Aayyush) [reply server/neptune/temporal/client.go:139]: Ahh okay!

**Review by Aayyush — APPROVED** [review]

**Review by samrabelachew — APPROVED** [review]

**Pull request merged**